server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -0,0 +1,403 @@
+            return Collections.emptyIterator();
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return Collections.emptyIterator();
+        }
+
+        if (value instanceof List<?> list) {
+            return (Iterator<Object>) list.iterator();
+        }
+        return Collections.singleton(value).iterator();
+    }
+
+    /**
+     * Get the value at this path, if there is no value then get the provided {@param defaultValue}
+     */
+    public Object get(Object defaultValue) {
+        if (leaf == null) {
+            return defaultValue;
+        }
+
+        return container.getOrDefault(leaf, defaultValue);
+    }
+
+    /**
+     * Get the value at the given index at this path or {@param defaultValue} if there is no such value.
+     */
+    public Object get(int index, Object defaultValue) {
+        if (leaf == null) {
+            return defaultValue;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value instanceof List<?> list) {
+            if (index < list.size()) {
+                return list.get(index);
+            }
+        } else if (value != MISSING && index == 0) {
+            return value;
+        }
+
+        return defaultValue;
+    }
+
+    /**
+     * Is there any value matching {@param predicate} at this path?
+     */
+    public boolean hasValue(Predicate<Object> predicate) {
+        if (leaf == null) {
+            return false;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return false;
+        }
+
+        if (value instanceof List<?> list) {
+            return list.stream().anyMatch(predicate);
+        }
+
+        return predicate.test(value);
+    }
+
+    // Value Update
+
+    /**
+     * Update each value at this path with the {@param transformer} {@link Function}.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField transform(Function<Object, Object> transformer) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            ((List<Object>) list).replaceAll(transformer::apply);
+        } else {
+            container.put(leaf, transformer.apply(value));
+        }
+
+        return this;
+    }
+
+    // Value Delete
+
+    /**
+     * Remove all duplicate values from this path.  List order is not preserved.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField deduplicate() {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            // Assume modifiable list
```

Comment:
Should this be an assert here, and perhaps related to my earlier comment, should we consider always making sure we have mutable list.

Reply:
An `assert` is going to be turned off in production.  The primary reason for using an `UnmodifiableList` is to error on modifications.

Reply:
Is it worth having a better error to let the user know what happened here as well?